Rework logging, second attempt

henryju opened the pull request.

henryju commented:
I experimented with the migration path for SonarJava: https://github.com/SonarSource/sonar-java/pull/4323/commits/ec9332a47457a4e56bdd7294f13ba4ca65c431f6

dbmeneses commented on plugin-api/src/main/java/org/sonar/api/utils/log/Loggers.java:
```diff
@@ -21,22 +21,14 @@
 
 public abstract class Loggers {
```
Since all the logging API in sonar-plugin-api will just wrap slf4j without providing any log interception for tests, couldn't we deprecate it and recommend everyone to move to using slf4j directly?

henryju replied:
That's my next plan :) but I wanted to keep the door open since we have the `setLevel` use case to sort out. So for now, no deprecation.

dbmeneses replied:
Ok. The javadoc mentions :The standard use-case is to customize logging of embedded 3rd-party libraries.". I guess that's a valid use case, and the plugin would have to get the logger for the 3rd party lib class to change its level.

But I wonder how much it's actually used. I couldn't find any uses of it in our internal plugins. It's used by SonarQube, but SonarQube and other products can do that directly with logback.

dbmeneses commented on plugin-api/src/main/java/org/sonar/api/utils/log/LoggersLoader.java:
```diff
@@ -19,24 +19,22 @@
-  protected Logger newInstance(String name) {
-    return new ConsoleLogger();
-  }
+  private static final Loggers INSTANCE = ServiceLoader.load(Loggers.class).findFirst().orElse(new Slf4jLoggers());
```
what's the intended use case for providing an alternative implementation of `Loggers`?  (less is more)

henryju replied:
In order to support the `setLevel`, which is not supported by slf4j. If we decide to drop this use case, or find a better way, we would be able to drop that.

dbmeneses commented on plugin-api/build.gradle:
```diff
@@ -1,6 +1,8 @@
 apply plugin: 'com.github.johnrengelman.shadow'
 
 dependencies {
+  api libs.slf4j
```
it doesn't matter that much because we don't use the new APIs in our wrappers, but could be a good chance to upgrade to slf4j 2?

henryju replied:
That's also in my next plans. That would require products to update to slf4j 2 when they will update to this new API version. I checked, and it is already the case in [SQ](https://github.com/SonarSource/sonar-enterprise/blob/d72ce6b2e755bb7b626add3dc19dbae49cf871cb/build.gradle#L335), but not yet in [SC](https://github.com/SonarSource/sonarcloud-core/blob/e9913df7419c72e79c30c9b8a3118cc0c537ea6d/build.gradle#L341) or SL. I have created a [ticket](https://sonarsource.atlassian.net/browse/SLCORE-455) in SL, and I will ask SC to do the same.

dbmeneses replied:
makes sense, thanks

dbmeneses commented on test-fixtures/build.gradle:
```diff
@@ -0,0 +1,51 @@
+dependencies {
```
almost feels like this should be forked out of the API into a testing library but it's probably not worth it :)

henryju replied:
This has always been a principle of the sonar-plugin-api to provide utility classes to help testing plugins. I think moving test utils to a separate artifact that won't be available in production is already a good step. :)

dbmeneses replied:
Agreed!

dbmeneses reviewed (approved):
Great cleanup, well done!

A few minor comments but overall LGTM

henryju commented on plugin-api/src/main/java/org/sonar/api/utils/log/Loggers.java:
```diff
@@ -21,14 +21,8 @@
-
   protected abstract Logger newInstance(String name);
 
-  protected abstract LoggerLevel getLevel();
```
This is protected, so not visible to plugins right?

dbmeneses replied:
missed that, that's correct.

henryju commented on plugin-api/build.gradle:
```diff
@@ -90,6 +94,13 @@ publishing {
       artifact sourcesJar
       artifact javadocJar
+      pom.withXml {
+        def dependenciesNode = asNode().appendNode('dependencies')
```
Indeed. The shadow plugin allows filtering out dependencies, but they still don't appear in the pom.
https://imperceptiblethoughts.com/shadow/publishing/#publishing-with-maven-publish-plugin
> For example, excluded dependencies are not automatically added to the POM file

henryju commented:
@dbmeneses I moved forward and removed the service loader. I have also marked the Sonar Logger facade as deprecated and added a changelog to help with migration.

henryju commented on test-fixtures/src/main/java/org/sonar/api/testfixtures/log/AbstractLogTester.java:
```diff
@@ -0,0 +1,162 @@
+  /**
+   * Logs with arguments in chronological order (item at index 0 is the oldest one)
+   */
+  public List<LogAndArguments> getLogs() {
```
I think this should be fixed by https://github.com/SonarSource/sonar-plugin-api/pull/85/commits/a900609e6afa002ae2b5e18e45b5623a0e9c20de

lukasz-jarocki-sonarsource replied:
Thank you and sorry I didn't notice.
